src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc

```diff
@@ -198,13 +198,38 @@ void FileWatcherCertificateProvider::ForceUpdate() {
       root_certificate_ = "";
     }
   }
+  absl::StatusOr<bool> keyCertMatch = false;
+  //check whether the present credentials are valid.
+  if (!pem_key_cert_pairs_.empty()) {
+    for (int i = 0; i < pem_key_cert_pairs_.size(); i++) {
+      keyCertMatch =
+          PrivateKeyAndCertificateMatch(pem_key_cert_pairs_.at(i).private_key(),
+                                        pem_key_cert_pairs_.at(i).cert_chain());
+      if (!(keyCertMatch.ok() && *keyCertMatch)) {
+        break;
+      }
+    }
+  }
   const bool identity_cert_changed =
       (!pem_key_cert_pairs.has_value() && !pem_key_cert_pairs_.empty()) ||
       (pem_key_cert_pairs.has_value() &&
        pem_key_cert_pairs_ != *pem_key_cert_pairs);
   if (identity_cert_changed) {
     if (pem_key_cert_pairs.has_value()) {
-      pem_key_cert_pairs_ = std::move(*pem_key_cert_pairs);
+      for (int i = 0; i < pem_key_cert_pairs->size(); i++) {
+        keyCertMatch = PrivateKeyAndCertificateMatch(
+          pem_key_cert_pairs->at(i).private_key(), pem_key_cert_pairs->at(i).cert_chain());
+        if (!(keyCertMatch.ok() && *keyCertMatch)) {
+          gpr_log(GPR_ERROR,
+                  "Certificate-key match result: %s",
+                  keyCertMatch.status().ToString().c_str());
+          break;
+        }
+      }
+      //in case of a match, update the credentials.
+      if (keyCertMatch.ok() && *keyCertMatch){
```

Comment:
Suggest using a new variable skip_reload (or maybe invalid_credential), which you set to true if PrivateKeyAndCertificateMatch fails. Then here we can check for if(!skip_reload).

Reply:
sounds good